torch/nn/modules/module.py

```diff
@@ -1914,21 +1918,16 @@ def state_dict(self, *args, destination=None, prefix='', keep_vars=False):
             if module is not None:
                 module.state_dict(destination=destination, prefix=prefix + name + '.', keep_vars=keep_vars)
         for hook in self._state_dict_hooks.values():
-            hook_result = hook(self, destination, prefix, local_metadata)
-            if hook_result is not None:
-                destination = hook_result
+            hook(self, destination, prefix, local_metadata)
```

Comment:
This fixes the issue pointed out in https://github.com/pytorch/pytorch/issues/117437

The semantic of "returning a new state_dict" for the hook is not well defined and only really made sense for the "root" module that `.state_dict()` was called from. Since each submodule takes in the root module's `destination` (i.e. state_dict) and adds keys into it, modifying the state_dict out of place in the hook does not make sense for a submodule, if we allow this behavior, it could potentially overwrite the root's destination without proper careful handling in the user's hook.

This is not BC-breaking since the API to register state_dict post hooks was private

Reply:
That sounds good. Do we have any user of the private API today that rely on this behavior that we know we are going to break though?

Reply:
Hm fair point, I am happy to just document this behavior instead!